compiler/vm/vmdef.nim

```diff
@@ -90,27 +91,224 @@ type
+      elementStride*: int ## The amount of bytes between the start of two consecutive 
+                         ## elements. Can be larger than the size of an element alone
+      elementType*: PVmType
+
+    sizeInBytes*: uint
+    alignment*: uint8 ## 1 shl `alignment` == real alignment value
+
+  CallableKind* = enum
+    ckDefault ## A referenc to a VM function
+    ckClosure
+    ckCallback ## A VmCallback
+    ckFFI
+    ckTemplate
+
+  # XXX: design is temporary
+  VmFunctionObject* = object ## An object storing a callable
+    case kind*: CallableKind
+    of ckDefault:
+      prcSym*: PNode # TODO: needs to be a `PSym`
+    of ckClosure:
+      closureEnv*: GlobalRef ## A reference to the closure environment
+      closureTarget*: int
+    of ckCallback:
+      cbOffset*: int
+    of ckFFI:
+      ffiPrcSym*: PSym
+    of ckTemplate:
+      templateSym*: PSym
+
+  # XXX: There are actually two kinds of location handles
+  # 1) those pointing to a 'real' atom
+  # 2) those pointing to the start a pseudo atom
+  # Maybe use the Nim's type system as a way to better document where which kinds
+  # is expected/used 
+
+  # XXX: once first class openArray support is available during for bootstrapping,
+  # `MemoryHandle` could be turned into `openArray[byte]`
+  MemoryHandle* = object
+    # XXX: without this dummy pointer, memory gets corrupted in ways
+    # that lead to random illegal storage accesses. The corruption very likely
+    # still happens but it seems to be more hidden. It's likely that some part
+    # of the VM writes past their the allocated memory range
+    _: pointer
+    p: ptr Atom
+    l: int
+  VmMemoryRange* = openArray[byte] # TODO: rename to `VmMemoryRegion`
+
+  TypedHandle* = tuple[h: MemoryHandle, typ: PVmType]
+
+  VmSeq* = object
+    length*: int
+    data*: MemoryHandle
+
+  VmString* = distinct VmSeq
+
+  VmSet* = object
+    # XXX maybe use `ptr IntSet`? Can't use IntSet with automatic storage duration since
+    # sizeof(IntSet) is also huge (smaller than set[uint16] however)
+
+    # XXX should `setElems` be ordered? Would speed up contains checks but require sorting on modification
+    # XXX `setElem` could also be run-length encoded for speed and memory efficiency...
+
+    # XXX In the end, VmSets should be embedded into their surrounding storage, without any heap allocations
+    elems* : seq[uint16]
+
+  Atom* {.union.} = object
+    intVal*: BiggestInt ## akInt
+    floatVal*: BiggestFloat ## akFloat
+    ptrVal*: ptr Atom ## akPtr
+    setVal*: VmSet ## akSet
+    strVal*: VmString ## akString
+    seqVal*: VmSeq ## akSeq
+    refVal*: GlobalRef ## akRef
+    callableVal*: VmFunctionObject ## akCallable
+
+    nodeVal*: PNode ## akPNode
+
+  # TODO: needs a better name
+  VmObjectAtom* = object ## A container that stores a value. Single atom
+                         ## values are stored in-place while complex values
+                         ## are seperately allocated 
+    case kind*: AtomKind
```

Comment:
This field alone add 7 bytes of unused memory due to alignment. I think an union and deciding how to interpret it based on the `typ` field is a much better solution.

In the future, if we move away from `PVmType` to an index-based solution (or you might already did that but I haven't read that far :stuck_out_tongue:), then accessing `typ` is pretty much guaranteed to be in cache ("free") due to the amount of type checks being done AFAICT.

Reply:
`VmObjectAtom` is obsolete. It's a leftover from a previous iteration that's still used for constants and few other things, but these will eventually use something different (already done in my local branch)

Reply:
How would an index-based solution work? The types being indices into a table that stores the actual layout information?

Reply:
Yep, it'd keep types close, very useful when you need to perform a lot of checks as types are more likely to be in cache.

A trick we could use in the future is to organize the type in a way such that subtypes can be computed as being in a given range (for `of` operator implementation).

Reply:
`PVmType` could be turned into `ptr VmType` and use a custom allocator (something like a bump pointer allocator) in order to keep types closer in memory. Of course, they still wouldn't be stored in a single continuous memory region.

The downside I see with the index-based approach is, that you'd have to carry around the table anywhere you'd want to access type information, which would make function interfaces more complex. But this would also make the type info lookup/access visible at the function interface, which I'd consider a plus.

Reply:
You can mitigate that downside by adding `TCtx` access everywhere :P It's the meat of the VM after all.